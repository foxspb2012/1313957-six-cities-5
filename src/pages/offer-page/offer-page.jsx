import React, {PureComponent} from 'react';
import {connect} from 'react-redux';
import * as Type from '../../prop-types';
import Map from '../../components/map/map';
import Header from '../../components/header/header';
import {Operation as OffersOperation} from '../../store/offers/offers';
import {Operation as ReviewsOperation} from '../../store/reviews/reviews';
import OfferList, {OffersListType} from '../../components/offer-list/offer-list';
import {getIsLoadedOffer, getIsLoadedNearOffers} from '../../store/offers/selectors';
import {getNearOffers, getMapOffers, getOneOffer} from '../../store/offers/selectors';
import OfferPageContent from '../../components/offer-page-content/offer-page-content';

class OfferPage extends PureComponent {
  componentDidMount() {
    this._update();
  }

  componentDidUpdate(prevProps) {
    if (prevProps.id !== this.props.id) {
      this._update();
    }
  }

  _update() {
    const {loadOneOffer, loadNearOffers, loadReviews} = this.props;
    loadOneOffer();
    loadNearOffers();
    loadReviews();
  }

  render() {
    const {mapOffers, nearOffers, offer, isLoadedOffer, isLoadedNearOffers} = this.props;
    const {location} = offer;

    return (
      <div className="page">
        <Header/>
        <main className="page__main page__main--property">
          <section className="property">
            <OfferPageContent
              offer={offer}
              isLoaded={isLoadedOffer}
            />
            <Map
              blockClassName={`property`}
              centerCoordinates={location}
              offers={mapOffers}
              isLoaded={isLoadedOffer && isLoadedNearOffers}
            />
          </section>
          <div className="container">
            <section className="near-places places">
              <h2 className="near-places__title">
                      Other places in the neighbourhood
              </h2>
              <OfferList
                typeClass={OffersListType.NEAR}
                offers={nearOffers}
                isLoaded={isLoadedNearOffers}
              />
            </section>
          </div>
        </main>
      </div>
    );
  }
}

OfferPage.propTypes = {
  id: Type.ID,
  mapOffers: Type.OFFERS_LIST,
  offer: Type.OFFER,
  isLoadedOffer: Type.FLAG,
  isLoadedNearOffers: Type.FLAG,
  nearOffers: Type.OFFERS_LIST,
  loadOneOffer: Type.FUNCTION,
  loadNearOffers: Type.FUNCTION,
  loadReviews: Type.FUNCTION,
  newId: Type.ID,
  oldId: Type.ID,
};

const mapStateToProps = (state) => ({
  mapOffers: getMapOffers(state),
  offer: getOneOffer(state),
  nearOffers: getNearOffers(state),
  isLoadedOffer: getIsLoadedOffer(state),
  isLoadedNearOffers: getIsLoadedNearOffers(state),
});

const mapDispatchToProps = (dispatch, {id}) => ({
  loadOneOffer: () => {
    dispatch(OffersOperation.loadOneOffer(id));
  },
  loadNearOffers: () => {
    dispatch(OffersOperation.loadNearOffers(id));
  },
  loadReviews: () => {
    dispatch(ReviewsOperation.loadReviews(id));
  },
});

export {OfferPage};
export default connect(mapStateToProps, mapDispatchToProps)(OfferPage);
