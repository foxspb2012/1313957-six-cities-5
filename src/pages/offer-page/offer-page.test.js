import React from 'react';
import history from '../../history';
import {Provider} from 'react-redux';
import {OfferPage} from './offer-page';
import {Router} from 'react-router-dom';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';

const NameSpace = {
  APP: `APP`,
  USER: `USER`,
  OFFERS: `OFFERS`,
  REVIEWS: `REVIEWS`,
};

const getOffersMock = (count) => {
  const templateOffers = Array(count)
    .fill(``)
    .map((_, index) => {
      return {
        'id': index + 1,
        'city': {
          'name': `Paris`,
          'location': {
            'coordinates': [48.85661, 2.351499],
            'zoom': 13
          }
        },
        'type': `room`,
        'name': `Wood and stone place`,
        'description': `An independent House, strategically located between Rembrand Square and National Opera, but where the bustle of the city comes to rest in this alley flowery and colorful.`,
        'photos': [`img/apartment-01.jpg`, `img/apartment-03.jpg`],
        'isFavorite': Boolean(Math.random()),
        'isPremium': Boolean(Math.random()),
        'rating': 1,
        'price': 120,
        'bedroomAmount': 1,
        'guestAmount': 1,
        'features': [`Dishwasher`, `Baby seat`],
        'host': {
          'name': `Host`,
          'avatar': `img/avatar-angelina.jpg`,
          'isPro': false,
          'id': index + 1
        },
        'location': {
          'coordinates': [48.87961000000001, 2.353499],
          'zoom': 16
        },
        'mainPhoto': `img/apartment-02.jpg`,
      };
    });

  if (count === 1) {
    const [oneOffer] = templateOffers;
    return oneOffer;
  }

  return templateOffers;
};

const mockStateOffersAndNearOffers = {
  [NameSpace.APP]: {
    activeCity: `Paris`,
    activeSortType: `popular`,
    activeOffer: null
  },
  [NameSpace.OFFERS]: {
    offers: [],
    offer: {},
    nearOffers: [],
    favoriteOffers: [],
    error: null,
    isLoaded: false,
    isLoadedOffer: true,
    isLoadedNearOffers: true,
    isLoadedFavoritesOffers: false,
  },
  [NameSpace.REVIEWS]: {
    reviews: [],
    isReviewSending: false,
    error: null
  },
  [NameSpace.USER]: {
    authorizationStatus: `UNAUTHORIZED`,
    email: `foxspb@yandex.ru`
  }
};

const testing = () => {};

describe(`Should OfferPage render correctly`, () => {
  it(`Should OfferPage render correctly with offer`, () => {
    const mockStore = configureStore();
    const store = mockStore(mockStateOffersAndNearOffers);
    const tree = renderer
      .create(
          <Provider store={store}>
            <Router history={history}>
              <OfferPage
                newId={1}
                offer={getOffersMock(1)}
                loadOneOffer={testing}
                loadNearOffers={testing}
                loadReviews={testing}
              />
            </Router>
          </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
