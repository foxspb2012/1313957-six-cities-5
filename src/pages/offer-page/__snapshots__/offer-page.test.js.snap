// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should OfferPage render correctly Should OfferPage render correctly with offer 1`] = `
<div
  className="page"
>
  <header
    className="header"
  >
    <div
      className="container"
    >
      <div
        className="header__wrapper"
      >
        <div
          className="header__left"
        >
          <a
            className="header__logo-link header__logo-link--active"
            href="/"
            onClick={[Function]}
          >
            <img
              alt="6 cities logo"
              className="header__logo"
              height="41"
              src="/img/logo.svg"
              width="81"
            />
          </a>
        </div>
        <nav
          className="header__nav"
        >
          <ul
            className="header__nav-list"
          >
            <li
              className="header__nav-item user"
            >
              <a
                className="header__nav-link header__nav-link--profile"
                href="/favorites"
                onClick={[Function]}
              >
                <div
                  className="header__avatar-wrapper user__avatar-wrapper"
                />
                <span
                  className="header__user-name user__name"
                >
                  foxspb@yandex.ru
                </span>
              </a>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </header>
  <main
    className="page__main page__main--property"
  >
    <section
      className="property"
    >
      <div
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "column",
            "fontSize": "38px",
            "fontStyle": "oblique",
            "fontWeight": 700,
            "justifyContent": "flex-start",
            "lineHeight": 0.7,
            "margin": "20px",
            "padding": "30px",
            "textAlign": "center",
            "width": "100%",
          }
        }
      >
        <h3
          style={
            Object {
              "alignItems": "center",
              "display": "flex",
              "justifyContent": "space-between",
            }
          }
        >
          <img
            alt="logo"
            height="50"
            src="/img/logo.svg"
            style={
              Object {
                "marginRight": "20px",
              }
            }
            width="100"
          />
          <svg
            height="100%"
            version="1.1"
            width="100%"
            xmlns="http://www.w3.org/2000/svg"
            xmlnsXlink="http://www.w3.org/1999/xlink"
          >
            <path
              id="path"
            >
              <animate
                attributeName="d"
                begin="0s"
                dur="3s"
                from="m0,110 h0"
                repeatCount="indefinite"
                to="m0,110 h1100"
              />
            </path>
            <text
              fill="black"
            >
              <textPath
                xlinkHref="#path"
              >
                Loading...
              </textPath>
            </text>
          </svg>
        </h3>
      </div>
      <div
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "column",
            "fontSize": "38px",
            "fontStyle": "oblique",
            "fontWeight": 700,
            "justifyContent": "flex-start",
            "lineHeight": 0.7,
            "margin": "20px",
            "padding": "30px",
            "textAlign": "center",
            "width": "100%",
          }
        }
      >
        <h3
          style={
            Object {
              "alignItems": "center",
              "display": "flex",
              "justifyContent": "space-between",
            }
          }
        >
          <img
            alt="logo"
            height="50"
            src="/img/logo.svg"
            style={
              Object {
                "marginRight": "20px",
              }
            }
            width="100"
          />
          <svg
            height="100%"
            version="1.1"
            width="100%"
            xmlns="http://www.w3.org/2000/svg"
            xmlnsXlink="http://www.w3.org/1999/xlink"
          >
            <path
              id="path"
            >
              <animate
                attributeName="d"
                begin="0s"
                dur="3s"
                from="m0,110 h0"
                repeatCount="indefinite"
                to="m0,110 h1100"
              />
            </path>
            <text
              fill="black"
            >
              <textPath
                xlinkHref="#path"
              >
                Loading...
              </textPath>
            </text>
          </svg>
        </h3>
      </div>
    </section>
    <div
      className="container"
    >
      <section
        className="near-places places"
      >
        <h2
          className="near-places__title"
        >
          Other places in the neighbourhood
        </h2>
        <div
          style={
            Object {
              "alignItems": "center",
              "display": "flex",
              "flexDirection": "column",
              "fontSize": "38px",
              "fontStyle": "oblique",
              "fontWeight": 700,
              "justifyContent": "flex-start",
              "lineHeight": 0.7,
              "margin": "20px",
              "padding": "30px",
              "textAlign": "center",
              "width": "100%",
            }
          }
        >
          <h3
            style={
              Object {
                "alignItems": "center",
                "display": "flex",
                "justifyContent": "space-between",
              }
            }
          >
            <img
              alt="logo"
              height="50"
              src="/img/logo.svg"
              style={
                Object {
                  "marginRight": "20px",
                }
              }
              width="100"
            />
            <svg
              height="100%"
              version="1.1"
              width="100%"
              xmlns="http://www.w3.org/2000/svg"
              xmlnsXlink="http://www.w3.org/1999/xlink"
            >
              <path
                id="path"
              >
                <animate
                  attributeName="d"
                  begin="0s"
                  dur="3s"
                  from="m0,110 h0"
                  repeatCount="indefinite"
                  to="m0,110 h1100"
                />
              </path>
              <text
                fill="black"
              >
                <textPath
                  xlinkHref="#path"
                >
                  Loading...
                </textPath>
              </text>
            </svg>
          </h3>
        </div>
      </section>
    </div>
  </main>
</div>
`;
